### PR TITLE
Added support for basic session based authentication

### PR DESCRIPTION
**Pull request by TMaYaD**
usage:

```terraform
locals {
  auth_json = {
    user = {
      login    = "admin"
      password = "supersecret"
    }
  }
}

provider "restapi" {
  uri                  = "https://MY_API_SERVER/api/"
  debug                = true
  insecure             = true
  debug = true
  session_auth {
    path         = "/sessions"
    data         = "${jsonencode(local.auth_json)}"
  }
```

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -3,7 +3,7 @@ package main
 import (
   "github.com/hashicorp/terraform/plugin"
   "github.com/hashicorp/terraform/terraform"
-  "github.com/Mastercard/terraform-provider-restapi/restapi"
+  "github.com/TMaYaD/terraform-provider-restapi/restapi"
```
Comment (DRuggeri) [inline main.go]: This will need to be fixed before merging

**Comment by DRuggeri** [comment]
I'm not sure I totally follow the objective as this seems very API implementation specific. Can you share some docs on what this is and how it should work?

**Comment by TMaYaD** [comment]
One of the APIs we are working with requires session based authentication. For this we need to login before we do pretty much anything else. This change supports that flow. 

This is limited to POST request in current implementation but making it configurable is trivial. However I do not think any one uses PUT or GET to login.

Another alternative auth implementation is exchanging login credentials for a token which should then be used in subsequent calls in headers or with basic auth. That requires more extensive support for parsing the response of login and setting values based on that. We do have one API that requires such flow but "for now" we settled on using external data source to fetch the token and set it on restapi.

Note that such workaround won't work here unless we can set cookies manually in the jar. We can probably do it similar to how we do headers right now, but I deemed this to be simpler solution to 'initialize' the client by making a post request with cookie jar on.

File: main.go (hunk 1 above)
Comment (TMaYaD) [reply main.go]: Done.

**Comment by DRuggeri** [comment]
Thanks, @TMaYaD, I think I understand now. I'm afraid this seems to be too implementation specific for this provider. I would encourage exploring use of the [external data source](https://www.terraform.io/docs/providers/external/data_source.html) to create the value you would use for the header. This has the benefit of supporting any arbitrary API requirements.

**Comment by TMaYaD** [comment]
That's a good middle-ground I think. For anyone ending up here in the future:

create_session.sh
```bash
#!/bin/sh

set -eux

usage() {
  # NOTE: Keep sorted for easy git merges
  cat << E_O_F
Usage:
  -a API Endpoint
  -p password
E_O_F
}

# NOTE: Keep sorted for easy git merges
while getopts "a:p:r:z:h?" opt; do
    case "$opt" in
    a)  API_ENDPOINT=${OPTARG}
        ;;
    p)  PASSWORD=${OPTARG}
        ;;
    h|\?)
        usage
        exit 0
        ;;
    esac
done

curl -k -D - -H "Content-Type: application/json" -X POST -d '{"user": {"login":"admin","password":"'$PASSWORD'"}}' $API_ENDPOINT/sessions | \
awk -F ";" 'BEGIN{printf "{ \"value\": \""; ORS=";";}/Set-Cookie/{print substr($1, 13)}END{ORS="";print "\"}"}'
```

main.tf

```terraform
data "external" "session" {
  program = [
    "${path.module}/create_session.sh",
    "-a",
    "https://${local.API_HOST}/api/",
    "-p",
    "${local.PASSWORD}",
  ]
}

provider "restapi" {
  uri                  = "https://${local.API_HOST}/api/"
  use_cookies          = true
  write_returns_object = true
  headers = {
    Cookie = data.external.session.result["value"]
  }

}
```

**Pull request closed without merging**